Merge Request - Added data import via file loading and dropping w/ average speaks/ranks

Benjamin-Marks opened the pull request:
Full List of Changes:
- Readme.md: Updated to describe process for file uploading
- forms.py: refactored "UploadFileForm" to "UploadBackupForm," added separate form for uploading excel sheets to import data
- judge_views.py: corrected spelling error in output string
- initialize_tourney.py: Now creates backup directory if it does not exist. This prevents initialize tourney from failing when run for the first time on a computer
- pairing_views.py: refactored "UploadFileForm" to "UploadBackupForm"
- team_views.py: Removed unused computation code
- navigation.html: Added sub-menu entries for file uploading and manual scratch entry
- thanks.html: Added section to display output for data import script
- views.py: Added methods for scratch entry and file importing
- import_judges.py: Adjusted old data-import script to work with the online imports. Imports judges, should catch user error
- import_rooms.py: Created new script to input rooms. Should catch user error.
- import_scratches.py: Started editing this script before deciding I didn't really want this functionality
- import_teams.py: Adjusted old data-import script to work with the online imports. Imports teams and debaters, should catch user error.
- tab_logic.py: Edited Tab logic to allow for dropping of non-checked-in teams with average speaks and ranks
  - For the first 'X' rounds of the tournament, No Show teams forfeit with average speaks and ranks
  - 'X' can be adjusting by modifying the admin "lenient_late" setting 
    - Setting this to 0 will fully restore previous logic
  - When computing byes, logic now prefers not to give the bye to a team with only lenient no shows unless there are no teams with a regular loss available
- database files: Edited database files to add an admin "lenient_late" setting with a value of 1
- urls.py: Added entries for file uploading and adding scratches
- requirements/setup.txt: Added required support for xlrd and xlwt libraries to do excel spreadsheet parsing

jolynch commented on mittab/apps/tab/team_views.py:
```diff
@@ -301,19 +301,6 @@ def tab_card(request, team_id):
         round_stats[i][6] = (round_stats[i][6][0] + round_stats[i-1][6][0],
                              round_stats[i][6][1] + round_stats[i-1][6][1])
-
-    totals = [[0,0],[0,0],[0,0]]
```
Good catch

jolynch commented on mittab/apps/tab/forms.py:
```diff
@@ -7,9 +7,15 @@
     file  = forms.FileField(label="Your Backup File")
 
+class UploadDataForm(forms.Form):
+    teamFile = forms.FileField(label="Teams Data File", required=False)
```
%s/teamFile/team_file

same for other two

jolynch commented on mittab/apps/tab/views.py:
```diff
@@ -248,4 +272,47 @@ def view_scratches(request):
+    if request.method == 'POST':
+      form = UploadDataForm(request.POST, request.FILES)
+      if form.is_valid():
+        judgeErrors = scratchErrors = teamErrors = []
```
nix the camelCase (sorry that some of the other files have it, a former contributor wrote a lot of java)

jolynch commented:
Sweet, these are great changes!

I'm going to take this review piece by piece since it touches so many files, sorry if the conversation log gets long.

 README.md :+1:
 mittab/apps/tab/forms.py : No camel case
 mittab/apps/tab/judge_views.py :+1: 
 mittab/apps/tab/management/commands/initialize_tourney.py :+1: 
 mittab/apps/tab/pairing_views.py :+1: 
 mittab/apps/tab/team_views.py :+1: Good catch
 mittab/apps/tab/templates/navigation.html :+1: 
 mittab/apps/tab/templates/thanks.html :+1: 
 mittab/apps/tab/views.py Have some comments, haven't written them down yet

I'll finish the rest of the review later today.

jolynch commented on mittab/apps/tab/views.py:
```diff
@@ -248,4 +272,47 @@ def view_scratches(request):
+        judgeErrors = scratchErrors = teamErrors = []
+        importName = ''
+        results = ''
+        #TODO: Could probably turn these checks into a loop
```
I'm ok with this as a TODO

jolynch commented on mittab/libs/data_import/import_judges.py:
```diff
@@ -18,43 +18,95 @@
+            judge_provider = ''
+
+        #iterate through schools until none are left
+        curCol = 4
```
nix the camelCase

jolynch commented on mittab/libs/data_import/import_teams.py:
```diff
@@ -18,72 +18,147 @@
+        deb1_provider = sh.cell(i,6).value
+
+
+        ironMan = False
```
nix the camelCase

jolynch commented on mittab/libs/data_import/import_teams.py:
```diff
@@ -18,72 +18,147 @@
+        except:
+            pass
+
+        team_school = sh.cell(i,1).value
```
nit: spaces between arguments, e.g. i,1 => i, 1
(I know we're inconsistent elsewhere, not a big deal if you want to be like "meh")

jolynch commented on mittab/libs/data_import/import_teams.py:
```diff
@@ -18,72 +18,147 @@
+        
+
+        #TODO: Verify there are not multiple free seeds from the same school
+        team_seed = sh.cell(i,2).value.lower()
```
probably want to .trim() this as well

jolynch commented on mittab/apps/tab/views.py:
```diff
@@ -31,6 +32,29 @@ def render_403(request):
+        team = form.cleaned_data['team'].name
+        return render_to_response('thanks.html', 
+                                  {'data_type': "Scratch",
+                                  'data_name': " from " + str(team) + " on " + str(judge),
```
" from {0} on {1}".format(team, judge)

jolynch commented on mittab/apps/tab/views.py:
```diff
@@ -248,4 +272,47 @@ def view_scratches(request):
+        importName = ''
+        results = ''
+        #TODO: Could probably turn these checks into a loop
+        if 'teamFile' in request.FILES:
```
So I'm a little confused how we get schools. I don't think I see where that automatically gets created from team/judge schools, but I might just be missing it.

Benjamin-Marks replied:
I was initially concerned about users having inconsistent capitalization of school names, especially when abbreviated (ie GTown vs Gtown), but I realized as I was responding Django allows you to check for this with name__iexact. Schools weren't added automatically before, but I'll update it.

jolynch commented on mittab/libs/data_import/import_scratches.py:
```diff
@@ -18,40 +18,38 @@
-    print "scratches entered"
+                scratch_errors.append[[j,t]]
+        except:
+            print "Epic fail"
```
Heh, maybe except Exception as e and then print it out? Sometimes helps me out when a team needs help if I can read the error log and see what's going on.

jolynch commented on mittab/libs/data_import/import_teams.py:
```diff
@@ -18,72 +18,147 @@
+        elif team_seed == 'unseeded' or team_seed == 'un' or team_seed == 'none' or team_seed == '':
+            team_seed = 0
+        else:
+            team_errors.append(team_name + ': Invalid Seed Value')
```
Maybe default to unseeded?

jolynch commented on mittab/libs/tab_logic.py:
```diff
@@ -619,13 +638,13 @@ def avg_deb_speaks(debater):
-    for returning average speaks or some such
-    """
-    return 0.0
+    if (TabSettings.get('lenient_late') >= round_number):
```
I'm curious about this implementation choice. My understanding of how most teams want to do avg speaks are when tab directors decide that a particular NoShow team deserves average speaks for a round they should get it. With this implementation, if I understand it correctly, we basically pick a round cutoff where if someone is noshow before that we've ok with it.

I'm curious if this is the right way to do it. On the one hand I like that a tab director can't be corrupt or something and selectively give good speaks to the teams they want, but on the other hand I think tab observers are kind of there specifically to prevent that kind of issue. On the other hand I can understand a situation where a tournament might want to be lenient for a team that traveled a really far way or something, but meh.

What are your thoughts?

Benjamin-Marks replied:
So the reason for this implementation is twofold. First, W&M's tab policy works this way (if you call before Round 1 is posted to tell us you'll be late, we drop you with avg speaks, as our tournament is so far south travel issues are common), I wasn't sure how other schools manage their late arrivals because I've never been affected by it at other tournaments. 

Secondly, this version of the implementation was much more easier to implement than a case-by-case basis, and I'm lazy. Doing case-by-case checks would require either substantially modifying the bye table or adding a separate table. The followup issue is that mit-tab can't predict when tab directors will justify noshows, so you either need individual prompts for each checked out team in every round, which seems excessive in the event that a team quits mid-tournament, or the tab director has to go into admin to manually add a team as being a lenient noshow. If there's a way to do case-by-case checks without tons of prompts, or you think having the prompts is fine, I'd be happy to re-do it.

jolynch replied:
So I'm thinking we either stick with this or alter the NoShow table to have an additional flag. I'm sorta ok with their being a default "lenient" round that we can change via TabSettings as long as the default is 0.

I think the big issue I've always had with lenient noshows is that having a hardline policy of "if you're late you're tabbed at 0" has been good for making people actually show up on time, but maybe I'm out of touch on that. I'm fine with it as an option anyhow.

jolynch commented:
mittab/libs/data_import/import_judges.py Minor syntax, :+1: 
 mittab/libs/data_import/import_rooms.py Same minor syntax issues, :+1: 
 mittab/libs/data_import/import_scratches.py, Minor change to exception handling :+1:
 mittab/libs/data_import/import_teams.py, Minor suggestions
mittab/libs/tab_logic.py Not sure about this choice of implementation, let's chat about it

jolynch commented:
Hm, I wonder why the tests are failing

jolynch commented on mittab/libs/tab_logic.py:
```diff
@@ -129,6 +129,25 @@ def pair_round():
                 if bracket == 0:
                     byeint = len(list_of_teams[bracket])-1
+                    #Don't give the bye to a team who only has lenient NoShows
+                    if (TabSettings.get('lenient_late') + 1 >= current_round):
```
I don't like this special exception from a code point of view or a tabulation point of view.

Code: We actually have code to "select" a bye twice, once in this part of the if block and again in the next section for dealing with 1-up teams getting the bye. If we keep this exception let's factor it out into a select_bye method or some such?

Tabulation: Byes are not a punishment (they're equivalent to a win by forfeit). If there are teams that have only had noshow loses, they should be just as eligible for a bye as anyone else.

jolynch commented on mittab/libs/tab_logic.py:
```diff
@@ -753,13 +773,13 @@ def avg_deb_ranks(debater):
-def debater_forfeit_ranks(debater):
-    """ Calculate a debater's ranks for a forfeit round
+def debater_forfeit_ranks(debater, round_number):
+    #Calculate a debater's ranks for a forfeit round
```
nit: space after #

jolynch commented:
Alright, so I think the tests are failing because lenient late defaults to 1. Can we make it default to 0 and see if that fixes things (alternatively dig into the fixture data and figure out how to change it).

@Benjamin-Marks This looks awesome, really excited to merge it once the tests pass!

Benjamin-Marks commented:
I've pushed one additional change: You're right, the tests were failing because of the lenient_late setting defaulting to 1. The failures persisted over a couple of commits because even though I had updated the clean database, the "make test" command doesn't initialize a new database. I've updated the Makefile so that running "make test" runs "initialize tourney" before doing the tests, so any changes made to the clean database are copied over and used in the test.

Otherwise I'm fine with everything you've mentioned and have updated it appropriately. I'm fine updating the NoShow table to have an additional flag in the future, but wanted to get this to a state where we could get it committed, and I'll look into those changes over the next couple of weeks.

Thanks for all of your advice with the code, happy I got the chance to help out!

jolynch commented:
@Benjamin-Marks Merged, I'll rebuild an image and start using it at tournaments asap.

Thanks for the contribution!
